src/v/cluster/archival/ntp_archiver_service.cc

```diff
@@ -1959,7 +1959,22 @@ ntp_archiver::wait_uploads_complete(
     result.checks_disabled
       = config::shard_local_cfg()
           .cloud_storage_disable_upload_consistency_checks.value();
-    if (!result.checks_disabled) {
+    auto skip_metadata_check = [this] {
+        // Special handling of the situation when the gap was created
+        // while the archiver was paused with 'redpanda.remote.allowgaps'
+        // set to 'true'.
+        // If the property is set to true and the local start offset
+        // doesn't match the last uploaded offset we should skip the
+        // check.
+        auto manifest_last = manifest().get_last_offset();
+        auto local_first = _parent.raft_start_offset();
+        model::offset manifest_next = model::next_offset(manifest_last);
+        bool gaps_allowed
+          = _parent.log()->config().is_remote_allow_gaps_enabled();
+        return gaps_allowed && !manifest().empty()
+               && manifest_next < local_first;
+    }();
+    if (!skip_metadata_check && !result.checks_disabled) {
```

Comment:
Do we think there would be value in performing the checks under the prior conditions (i.e. only `!result.checks_disabled`) and then using `skip_metadata_check` to gate some error logging if our assumptions are incorrect? I've not thought through what that would mean exactly, but it's driven by an observation that the checks are now only performed if two conditions are met, but one of those conditions, namely the calculation of `skip_metadata_check` feels a little fragile with all the offset checking. I'm just wondering if we should be conservative with the correctness of `skip_metadata_check` and then evolve it to this current version once we gain confidence?

Reply:
I think there is no need to be conservative with the check because we have an escape hatch. And I don't think there is a roof for incorrect behavior here. We do the metadata check if the metadata check not disabled via the escape hatch and if we're not resuming uploads with inconsistency. The last case is only true if the gaps were allowed by the user and there is a gap between local data and data in the cloud. The invariant is that the first segment upload will disable this case for the future because of the way 'max_collectible_offset' works.

Reply:
> I think there is no need to be conservative with the check because we have an escape hatch. And I don't think there is a roof for incorrect behavior here. We do the metadata check if the metadata check not disabled via the escape hatch and if we're not resuming uploads with inconsistency. The last case is only true if the gaps were allowed by the user and there is a gap between local data and data in the cloud. The invariant is that the first segment upload will disable this case for the future because of the way 'max_collectible_offset' works.

sgtm. lets make sure we have test coverage of these scenarios